[Core] Assertion error when refreshing status for INIT cluster

To reproduce:
```
sky launch -c existing-up-cluster
# ctrl-c when it is launching
sky status -r existing-up-cluster
sky launch -c existing-up-cluster
```

```
sky/backends/backend_utils.py", line 2024, in run_ray_status_to_check_ray_cluster_healthy
    assert handle.head_ssh_port is not None, handle
AssertionError: ResourceHandle(
        cluster_name=spot-dev-zhwu,
        cluster_name_on_cloud=spot-dev-zhwu,
        head_ip=34.135.151.192,
        stable_internal_external_ips=[('10.128.0.68', '34.135.151.192')],
        stable_ssh_ports=None,
        cluster_yaml=/Users/zhwu/.sky/generated/spot-dev-zhwu.yml, 
        launched_resources=1x GCP(n2-standard-8, image_id={'us-central1': 'projects/skypilot-managed-spot/global/machineImages/spot-dev-image'}), 
        docker_user=None,
        tpu_create_script=None, 
        tpu_delete_script=None)
```

**This is a serious problem, as once the cluster enter this state, none of our clis can operate on the cluster to get it out of the state.**

--- a/sky/backends/backend_utils.py
+++ b/sky/backends/backend_utils.py
@@ -2021,6 +2021,9 @@ def _update_cluster_status_no_lock(
             # Check if ray cluster status is healthy.
             ssh_credentials = ssh_credential_from_yaml(handle.cluster_yaml,
                                                        handle.docker_user)
+            # Potentially, refresh the ports if they are not cached in handle.
+            # It is ok, as the refreshing of ports is fast.
+            handle.external_ssh_ports()
             assert handle.head_ssh_port is not None, handle
             runner = command_runner.SSHCommandRunner(external_ips[0],
                                                      **ssh_credentials,

--- a/sky/backends/backend_utils.py
+++ b/sky/backends/backend_utils.py
@@ -2016,13 +2016,11 @@ def _update_cluster_status_no_lock(
                 raise exceptions.FetchIPError(
                     reason=exceptions.FetchIPError.Reason.HEAD)
 
-            if handle.head_ssh_port is None:
-                # Refresh the ssh ports. It is ok to refresh as it is fast.
-                handle.external_ssh_ports()
-                if handle.head_ssh_port is None:
-                    raise RuntimeError(
-                        f'Refreshing status ({cluster_name!r}): failed '
-                        f'to get the ssh ports. Handle: {handle}')
+            # Potentially refresh the external SSH ports, in case the existing
+            # cluster before #2491 was launched without external SSH ports
+            # cached.
+            external_ssh_ports = handle.external_ssh_ports()
+            head_ssh_port = external_ssh_ports[0]
 
             # Check if ray cluster status is healthy.
             ssh_credentials = ssh_credential_from_yaml(handle.cluster_yaml,
@@ -2030,7 +2028,7 @@ def _update_cluster_status_no_lock(
 
             runner = command_runner.SSHCommandRunner(external_ips[0],
                                                      **ssh_credentials,
-                                                     port=handle.head_ssh_port)
+                                                     port=head_ssh_port)
             rc, output, stderr = runner.run(
                 RAY_STATUS_WITH_SKY_RAY_PORT_COMMAND,
                 stream_logs=False,
@@ -2055,6 +2053,12 @@ def _update_cluster_status_no_lock(
                 f'Refreshing status ({cluster_name!r}) failed to get IPs.')
         except RuntimeError as e:
             logger.debug(str(e))
+        except Exception as e:  # pylint: disable=broad-except
+            # This can be raised by `external_ssh_ports()`, due to the
+            # underlying call to kubernetes API.
+            logger.debug(
+                f'Refreshing status ({cluster_name!r}) failed: '
+                f'{common_utils.format_exception(e, use_bracket=True)}')
         return False
 
     # Determining if the cluster is healthy (UP):

--- a/sky/backends/backend_utils.py
+++ b/sky/backends/backend_utils.py
@@ -2014,17 +2014,24 @@ def _update_cluster_status_no_lock(
             # triggered.
             if external_ips is None or len(external_ips) == 0:
                 logger.debug(f'Refreshing status ({cluster_name!r}): No cached '
-                             f'IPs found. External IPs: {external_ips}')
+                             f'IPs found. Handle: {handle}')
                 raise exceptions.FetchIPError(
                     reason=exceptions.FetchIPError.Reason.HEAD)
+
+            if handle.head_ssh_port is None:
+                # Refresh the ssh ports. It is ok to refresh as it is fast.
+                handle.external_ssh_ports()
+                if handle.head_ssh_port is None:
+                    logger.debug(
+                        f'Refreshing status ({cluster_name!r}): failed '
+                        f'to get the ssh ports. Handle: {handle}')
+                    raise exceptions.FetchIPError(
+                        reason=exceptions.FetchIPError.Reason.HEAD)
 
             # Check if ray cluster status is healthy.
             ssh_credentials = ssh_credential_from_yaml(handle.cluster_yaml,
                                                        handle.docker_user)
-            # Potentially, refresh the ports if they are not cached in handle.
-            # It is ok, as the refreshing of ports is fast.
-            handle.external_ssh_ports()
-            assert handle.head_ssh_port is not None, handle
+
             runner = command_runner.SSHCommandRunner(external_ips[0],
                                                      **ssh_credentials,
                                                      port=handle.head_ssh_port)

--- a/sky/backends/backend_utils.py
+++ b/sky/backends/backend_utils.py
@@ -2022,11 +2022,9 @@ def _update_cluster_status_no_lock(
                 # Refresh the ssh ports. It is ok to refresh as it is fast.
                 handle.external_ssh_ports()
                 if handle.head_ssh_port is None:
-                    logger.debug(
+                    raise RuntimeError(
                         f'Refreshing status ({cluster_name!r}): failed '
                         f'to get the ssh ports. Handle: {handle}')
-                    raise exceptions.FetchIPError(
-                        reason=exceptions.FetchIPError.Reason.HEAD)
 
             # Check if ray cluster status is healthy.
             ssh_credentials = ssh_credential_from_yaml(handle.cluster_yaml,

--- a/sky/backends/cloud_vm_ray_backend.py
+++ b/sky/backends/cloud_vm_ray_backend.py
@@ -1547,7 +1547,7 @@ class RetryingVmProvisioner(object):
                 tpu_delete_script=config_dict.get('tpu-delete-script'),
                 # Use the previous cluster's IPs and ports if available to
                 # optimize the case where the cluster is restarted, i.e., no
-                # need to query the IPs from the cloud provider.
+                # need to query IPs and ports from the cloud provider.
                 stable_internal_external_ips=prev_cluster_ips,
                 stable_ssh_ports=prev_ssh_ports)
             usage_lib.messages.usage.update_final_cluster_status(
@@ -2317,6 +2317,7 @@ class CloudVmRayResourceHandle(backends.backend.ResourceHandle):
         self.launched_resources = self.launched_resources.copy(region=region)
 
     def update_ssh_ports(self, max_attempts: int = 1) -> None:
+        """Updates the cluster SSH ports cached in the handle."""
         # TODO(romilb): Replace this with a call to the cloud class to get ports
         # Use port 22 for everything except Kubernetes
         if not isinstance(self.launched_resources.cloud, clouds.Kubernetes):
@@ -2364,6 +2365,10 @@ class CloudVmRayResourceHandle(backends.backend.ResourceHandle):
             external_ips: The external IPs to use for the cluster. Similar to
                 internal_ips, it is an optimization to avoid retrieving the
                 external IPs from the cloud provider.
+
+        Raises:
+            exceptions.FetchIPError: if we failed to get the IPs. e.reason is
+                HEAD or WORKER.
         """
 
         def is_provided_ips_valid(ips: Optional[List[Optional[str]]]) -> bool:

--- a/sky/backends/cloud_vm_ray_backend.py
+++ b/sky/backends/cloud_vm_ray_backend.py
@@ -1524,9 +1524,10 @@ class RetryingVmProvisioner(object):
             if zones and len(zones) == 1:
                 launched_resources = launched_resources.copy(zone=zones[0].name)
 
-            prev_cluster_ips = None
+            prev_cluster_ips, prev_ssh_ports = None, None
             if prev_handle is not None:
                 prev_cluster_ips = prev_handle.stable_internal_external_ips
+                prev_ssh_ports = prev_handle.stable_ssh_ports
             # Record early, so if anything goes wrong, 'sky status' will show
             # the cluster name and users can appropriately 'sky down'.  It also
             # means a second 'sky launch -c <name>' will attempt to reuse.
@@ -1543,10 +1544,11 @@ class RetryingVmProvisioner(object):
                 launched_resources=launched_resources,
                 tpu_create_script=config_dict.get('tpu-create-script'),
                 tpu_delete_script=config_dict.get('tpu-delete-script'),
-                # Use the previous cluster's IPs if available to optimize
-                # the case where the cluster is restarted, i.e., no need to
-                # query the IPs from the cloud provider.
-                stable_internal_external_ips=prev_cluster_ips)
+                # Use the previous cluster's IPs and ports if available to
+                # optimize the case where the cluster is restarted, i.e., no
+                # need to query the IPs from the cloud provider.
+                stable_internal_external_ips=prev_cluster_ips,
+                stable_ssh_ports=prev_ssh_ports)
             usage_lib.messages.usage.update_final_cluster_status(
                 status_lib.ClusterStatus.INIT)
 

